src/detect-dns-response.c

```diff
@@ -0,0 +1,307 @@
+                }
+                break;
+            case DNS_RESPONSE_ADDITIONAL:
+                if (!SCDnsTxGetAdditionalRdata(txv, cbdata->response_id, &data, &data_len)) {
+                    InspectionBufferSetupMultiEmpty(buffer);
+                    return NULL;
+                }
+                break;
+            default:
+                InspectionBufferSetupMultiEmpty(buffer);
+                return NULL;
+        }
+    } else {
+        /* Get name values. */
+        switch (cbdata->response_section) {
+            case DNS_RESPONSE_QUERY:
+                if (!SCDnsTxGetQueryName(txv, true, cbdata->response_id, &data, &data_len)) {
+                    InspectionBufferSetupMultiEmpty(buffer);
+                    return NULL;
+                }
+                break;
+            case DNS_RESPONSE_ANSWER:
+                if (!SCDnsTxGetAnswerName(txv, true, cbdata->response_id, &data, &data_len)) {
+                    InspectionBufferSetupMultiEmpty(buffer);
+                    return NULL;
+                }
+                break;
+            case DNS_RESPONSE_AUTHORITY:
+                if (!SCDnsTxGetAuthorityName(txv, cbdata->response_id, &data, &data_len)) {
+                    InspectionBufferSetupMultiEmpty(buffer);
+                    return NULL;
+                }
+                break;
+            case DNS_RESPONSE_ADDITIONAL:
+                if (!SCDnsTxGetAdditionalName(txv, cbdata->response_id, &data, &data_len)) {
+                    InspectionBufferSetupMultiEmpty(buffer);
+                    return NULL;
+                }
+                break;
+            default:
+                InspectionBufferSetupMultiEmpty(buffer);
+                return NULL;
+        }
+    }
+
+    InspectionBufferSetupMulti(buffer, transforms, data, data_len);
+    buffer->flags = DETECT_CI_FLAGS_SINGLE;
+    return buffer;
+}
+
+static uint8_t DetectEngineInspectCb(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const struct DetectEngineAppInspectionEngine_ *engine, const Signature *s, Flow *f,
+        uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+{
+    const DetectEngineTransforms *transforms = NULL;
+    if (!engine->mpm) {
+        transforms = engine->v2.transforms;
+    }
+
+    uint32_t local_id = 0;
+    /* loop through each possible DNS response section */
+    for (uint8_t section = DNS_RESPONSE_QUERY; section < DNS_RESPONSE_MAX; section++) {
+        uint32_t response_id = 0;
+        /* loop through each record in section inspecting "name" and "rdata" */
+        while (1) {
+            struct DnsResponseGetDataArgs cbdata = { section, response_id, local_id };
+
+            /* do inspection for resource record "name" */
+            InspectionBuffer *buffer =
+                    GetBuffer(det_ctx, flags, transforms, txv, &cbdata, engine->sm_list, false);
+            if (buffer == NULL || buffer->inspect == NULL) {
+                local_id++;
```

Comment:
Do we need `local_id++` if we break right after ?

Reply:
Doesn't look like it.

Reply:
Nevermind, we're only breaking the inner loop, its still needed by the outer loop.